File: leetcode/medium/55-jump_game.ts
function canJump(nums: number[]): boolean {
    const memo = new Map<number, boolean>();

    function recursive(index: number): boolean {
        if (memo.has(index)) return memo.get(index) ?? false;
        if (index == nums.length - 1) return true;
        let found = false;
        for (var i = nums[index]; i > 0; i--) {
            const target = index + i;
            if (recursive(target)) {
                found = true;
                break;
            }
        }
        memo.set(index, found);
        return found;
    }

    return recursive(0);
};

File: leetcode/medium/55-jump_game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int max = nums[0];

        for (int i = 0; i < nums.size(); i++) {
            if (max >= nums.size() - 1) return true;
            if (nums[i] == 0 && max == i) return false;
            if (i + nums[i] > max) max = i + nums[i];
        }
        return true;        
    }
};
